### PR TITLE
Making changes to better support service worker registration as input

### LINKED ISSUE #30: Service Worker Registration
cc @jeffposnick @wibblymat @paullewis @petele

Me and Pete were discussion how to use the Propel library and one of the things that came up was how to use it with your own SW.

This isn't obvious and passing in a worker url and scope for something you've registered already for different reasons feels a bit weird to give the url and scope for.

The API at present is:

    new Client({workerUrl: '/sw.js', scope: './'});

Would anyone object to changing this to something a bit more permanent of:

    new Client('/sw.js'); // Scope is predefined as current script path + /.goog.propel.push

    new Client('./sw.js', './');

    new Client(swRegistration); 

Service work url can be obtained from the installing, waiting or active service worker object and the scope can be retrieved from the registration object.

Is there any downsides / edge cases to this?

The only reason I'm proposing ditching the options object is the worker.js file is currently empty and I'm not convinced shipping a default service worker is a good idea here, shipping with a library to help generate notifications makes sense, but it will still require some work on the developers behalf and expecting worker.js to exist in the same place as the client.js file feels a bit risky / unpredictable. The above at least forces some decision to be made.

What do people think? This is new and is likely to come up again with any libraries around background sync.

### PR DESCRIPTION
**Pull request by gauntface**
Fixes #30

@wibblymat @jeffposnick @petele 

This change should implement the changes we discussed in #30 regarding requiring a worker URL or Registration in the constructor.

I've added a constant for error messages so they are easier to handle in the tests with changes, although still not easily shareable by just importing a file because the push-client is transpiled the tests in the browser are not (so no easy way I can think of to require in a file) - Thoughts on this would be super helpful.

### CODE REVIEW

**Comment by gauntface** [comment]
FYI Tests are failing but seems to be a problem with Travis VM + Google Chrome Install (i.e. no browser to run the tests in).

File: src/client/push-client.js [hunk 1]
```diff
@@ -16,24 +16,24 @@ import SubscriptionFailedError from './subscription-failed-error';
...
 
+const ERROR_MESSAGES = {
+  'bad constructor': 'The PushClient constructor expects either service ' +
+    'worker registration or the path to a service worker file and an ' +
+    'optional scope string.',
+  'redundant worker': 'Worker became redundant'
+};
+
 const registrationReady = function(registration) {
   if (registration.active) {
```
Comment (ithinkihaveacat) [inline src/client/push-client.js]: It's not part of this PR, but I think there might be an issue here if the SW the code is waiting for is installing, but a previous one is active. In this case `registerReady()` will resolve immediately, even though the correct SW isn't actually available yet. Maybe check if there's anything `.installing` or `.waiting` before resolving, see http://stackoverflow.com/a/35705984/11543.

Comment (gauntface) [inline src/client/push-client.js]: This library only cares that the user can be subscribed / unsubscribed. Which service worker shouldn't matter but the subscription will transfer (or least it should do).

Is there any scenario where the latest service is required?

Comment (ithinkihaveacat) [inline src/client/push-client.js]: This has been discussed offline, but for the record, the situation I was thinking about was where the old SW doesn't handle `push` events, but the new one does. This could lead to messages being dropped until the new SW is controlling the page. (For example, it's plausible the server will attempt to send a message like "Thanks for subscribing" or "Here's what a notification looks like" as soon as it receives subscription information from the client--the messages might not get through.)

(I think the conclusion was that there's no elegant way to deal with this--it just needs to be documented that some combinations of SW/client states may be problematic.)

**Comment by wibblymat** [comment]
:+1:
